mm2src/mm2_main/src/rpc/lp_commands/lr_swap/lr_impl.rs

```diff
@@ -0,0 +1,350 @@
+    src_contract: Option<EthAddress>,
+    /// Source token amount in wei
+    src_amount: Option<U256>,
+    /// Source token decimals
+    src_decimals: Option<u8>,
+    /// Destination token contract address (to do interim LR swap into)
+    dst_contract: Option<EthAddress>,
+    /// Destination token amount in wei
+    dst_amount: Option<U256>,
+    /// Destination token decimals
+    dst_decimals: Option<u8>,
+    /// Chain id where interim LR swap occurs (obtained from the destination token)
+    chain_id: Option<u64>,
+    /// Estimated src token / dst token price
+    lr_price: Option<MmNumber>,
+    /// A quote from LR provider with tx data to do interim LR swap of src token to dst token
+    lr_swap_data: Option<ClassicSwapData>,
+}
+
+impl LrData {
+    #[allow(clippy::result_large_err)]
+    fn get_chain_contract_info(&self) -> MmResult<(String, String, u64), ApiIntegrationRpcError> {
+        let src_contract = self
+            .src_contract
+            .as_ref()
+            .ok_or(ApiIntegrationRpcError::InternalError(
+                "Source LR contract not set".to_owned(),
+            ))?
+            .display_address();
+        let dst_contract = self
+            .dst_contract
+            .as_ref()
+            .ok_or(ApiIntegrationRpcError::InternalError(
+                "Destination LR contract not set".to_owned(),
+            ))?
+            .display_address();
+        let chain_id = self
+            .chain_id
+            .ok_or(ApiIntegrationRpcError::InternalError("LR chain id not set".to_owned()))?;
+        Ok((src_contract, dst_contract, chain_id))
+    }
+}
+
+struct LrDataMap {
+    /// Map to store data needed for best price estimations for swaps with LR,
+    /// the key is the source and destination token pair from the LR swap part
+    inner: HashMap<(Ticker, Ticker), LrData>,
+}
+
+impl LrDataMap {
+    /// Init LR data map from the source token (mytoken) and tokens from orders
+    fn new_with_src_token(src_token: Ticker, orders: Vec<RpcOrderbookEntryV2>) -> Self {
+        Self {
+            inner: orders
+                .into_iter()
+                .map(|order| {
+                    ((src_token.clone(), order.coin.clone()), LrData {
+                        order,
+                        src_contract: None,
+                        src_decimals: None,
+                        src_amount: None,
+                        dst_contract: None,
+                        dst_amount: None,
+                        dst_decimals: None,
+                        chain_id: None,
+                        lr_price: None,
+                        lr_swap_data: None,
+                    })
+                })
+                .collect::<HashMap<(_, _), _>>(),
+        }
+    }
+
+    /// Calculate amounts of destination tokens required to fill ask orders for the requested base_amount:
+    /// multiplies base_amount by the order price (base_amount must be in 'coins')
```

Comment:
`base_amount must be in 'coins'`? what does that mean?
did u mean `base.decimals()`?

Reply:
I updated the comment: I mean 'coin units' (i.e. eth not wei)